BuildNewActor service

awead opened the pull request:
To support API integrations, we need to be able to build new or existing Actor objects given an identifier. This way, a client can provide either a Penn State access id or an Orcid id, and we'll be able to create the required Actor without any additional information.

This pattern assumes the client knows the id and that it is valid; otherwise, an error will (ultimately) be returned to them.

Fixes #910

awead committed:
BuildNewActor service

To support API integrations, we need to be able to build new or existing
Actor objects given an identifier. This way, a client can provide either
a Penn State access id or an Orcid id, and we'll be able to create the
required Actor without any additional information.

This pattern assumes the client knows the id and that it is valid;
otherwise, an error will (ultimately) be returned to them.

rschenk commented on app/models/actor.rb:
```diff
@@ -66,11 +64,9 @@ class Actor < ApplicationRecord
+            },
             presence: true,
-            on: :from_user
+            unless: -> { psu_id.present? }
```
I'm not sure this `unless psu_id.present?` is entirely in the right place. As I'm reading this, the intended behavior is, "orcid is required if psu_id is missing". This accomplishes that goal, but it _also_ affects the format regex too. I believe what will happen with this change is that, if psu_id is present, then I can type anything I want into the orcid field and its format will be accepted.

rschenk replied:
I think what we want is

```ruby
  validates :orcid,
            uniqueness: {
              case_sensitive: false,
              allow_nil: true
            },
            format: {
              with: /\A\d{15,15}[0-9X]{1,1}\z/,
              message: 'must be valid',
              allow_nil: true
            }
  validates :orcid,
            presence: true,
            unless: -> { psu_id.present? }
```

awead replied:
Yep. That makes sense to me.

awead replied:
@rschenk I've updated this and added a test case to reflect it.

awead commented on app/services/build_new_actor.rb:
```diff
@@ -0,0 +1,41 @@
+        build_with_psu_id(psu_id)
+      elsif orcid.present?
+        build_with_orcid(orcid)
+      else
```
There's another case I missed here: what if _both_ are present. The Penn State id would take precedence, but I'm not sure we would want that.

rschenk commented on app/models/actor.rb:
```diff
@@ -51,11 +51,9 @@ class Actor < ApplicationRecord
+            },
             presence: true,
-            on: :from_omniauth
+            unless: -> { orcid.present? }
```
We have the same situation here as we did on `orcid`. I think we need to separate to the uniqueness validation to run all the time, and the presence validation to run `unless orcid.present?`

awead replied:
yep, sure enough!

awead committed:
Raise NotFound error for PennState::SearchService

To align more with our Orcid service, PennState::SearchService should
raise errors for 404 Not Found responses. This necessitates rescuing
these errors as needed, but more is more consistent and avoids the "nil
obsession" code smell.
